run-tests.sh

```diff
@@ -14,14 +51,39 @@ do
   full_second_level_dir=$(echo "$dir" | cut -d/ -f1-2)
 
   # Only run if change happened at second level, since first level-changes don't require the tests to be re-run.
+  # If change happened at first level, $second_level_dir will be empty.
   if [ ! -z "$second_level_dir" -a "$second_level_dir" != " " ]; then
     if [[ ! " ${tested_dirs[*]} " =~ " ${full_second_level_dir} " ]]; then
+      echo "Executing tests for $full_second_level_dir"
+
       tested_dirs+=($full_second_level_dir)
       if [ -e $full_second_level_dir/requirements.txt ]; then
-        python -m pip install -r $full_second_level_dir/requirements.txt --extra-index-url https://download.pytorch.org/whl/cpu
+        $python_interpreter -m pip -q install -r $full_second_level_dir/requirements.txt --extra-index-url https://download.pytorch.org/whl/cpu --no-warn-script-location
+      fi
+
+      # Ensure proper spaCy version is installed.
+      spacy_version=$(
+        $python_interpreter -c "import srsly; data = srsly.read_yaml('${full_second_level_dir}/project.yml'); print(data.get('spacy_version', ''))"
+      )
+      if [ ! -z "$spacy_version" ]; then
+        $python_interpreter -m pip -q install "spacy${spacy_version}" --force-reinstall --no-warn-script-location
+      fi
+
+      $python_interpreter -m pytest -q -s $full_second_level_dir
+
+      # Mark as failure if exit code isn't either 0 (success) or 5 (no tests found).
+      if [[ $? != @(0|5) ]]; then
+        status=1
+      fi
+
+      if [ -e $full_second_level_dir/requirements.txt ]; then
+        $python_interpreter -m pip freeze --exclude torch --exclude wheel cupy-cuda111 srsly aiohttp > installed.txt
```

Comment:
If it's always reinstalling the dev requirements below, then I think the excludes can be limited to just the large things that we don't want to wipe out:

```suggestion
        $python_interpreter -m pip freeze --exclude torch --exclude cupy-cuda111 > installed.txt
```

`pip freeze` also already automatically excludes `setuptools, pip, distribute, wheel` unless you specify `--all`.